File: DSA/src/Questions/Trie/ImplementTrie2.java
package Questions.Trie;

/*
https://leetcode.com/problems/implement-trie-ii-prefix-tree/description/
 A trie (pronounced as "try") or prefix tree is a tree data structure used to efficiently store and retrieve keys in a dataset of strings. There are various applications of this data structure, such as autocomplete and spellchecker.

Implement the Trie class:

Trie() Initializes the trie object.
void insert(String word) Inserts the string word into the trie.
int countWordsEqualTo(String word) Returns the number of instances of the string word in the trie.
int countWordsStartingWith(String prefix) Returns the number of strings in the trie that have the string prefix as a prefix.
void erase(String word) Erases the string word from the trie.
 

Example 1:

Input
["Trie", "insert", "insert", "countWordsEqualTo", "countWordsStartingWith", "erase", "countWordsEqualTo", "countWordsStartingWith", "erase", "countWordsStartingWith"]
[[], ["apple"], ["apple"], ["apple"], ["app"], ["apple"], ["apple"], ["app"], ["apple"], ["app"]]
Output
[null, null, null, 2, 2, null, 1, 1, null, 0]

Explanation
Trie trie = new Trie();
trie.insert("apple");               // Inserts "apple".
trie.insert("apple");               // Inserts another "apple".
trie.countWordsEqualTo("apple");    // There are two instances of "apple" so return 2.
trie.countWordsStartingWith("app"); // "app" is a prefix of "apple" so return 2.
trie.erase("apple");                // Erases one "apple".
trie.countWordsEqualTo("apple");    // Now there is only one instance of "apple" so return 1.
trie.countWordsStartingWith("app"); // return 1
trie.erase("apple");                // Erases "apple". Now the trie is empty.
trie.countWordsStartingWith("app"); // return 0
 

Constraints:

1 <= word.length, prefix.length <= 2000
word and prefix consist only of lowercase English letters.
At most 3 * 104 calls in total will be made to insert, countWordsEqualTo, countWordsStartingWith, and erase.
It is guaranteed that for any function call to erase, the string word will exist in the trie.
insert: O(n)/O(n)
countWordsEqualTo: O(n)/O(1)
countWordsStartingWith: O(n)/O(1)
erase: O(n)/O(1)
 */
public class ImplementTrie2 {
    TrieNode root = new TrieNode();

    public void insert(String word) {
        TrieNode node = root;
        for (char c : word.toCharArray()) {
            int index = c - 'a';
            if (node.children[index] == null) {
                node.children[index] = new TrieNode();
            }
            node = node.children[index];
            node.pCount++;
        }
        node.count++;
    }

    public int countWordsEqualTo(String word) {
        TrieNode node = search(word);
        return node == null ? 0 : node.count;
    }

    public int countWordsStartingWith(String prefix) {
        TrieNode node = search(prefix);
        return node == null ? 0 : node.pCount;
    }

    public void erase(String word) {
        TrieNode node = root;
        for (char c : word.toCharArray()) {
            int index = c - 'a';
            node = node.children[index];
            node.pCount--;
        }
        node.count--;
    }

    private TrieNode search(String s) {
        TrieNode node = root;
        for (char c : s.toCharArray()) {
            int index = c - 'a';
            if (node.children[index] == null) {
                return null;
            }
            node = node.children[index];
        }
        return node;
    }

    public static void main(String[] args) {
        ImplementTrie2 trie = new ImplementTrie2();
        trie.insert("apple");
        trie.insert("apple");
        System.out.println(trie.countWordsEqualTo("apple"));
        System.out.println(trie.countWordsStartingWith("app"));
        trie.erase("apple");
        System.out.println(trie.countWordsEqualTo("apple"));
        System.out.println(trie.countWordsStartingWith("app"));
        trie.erase("apple");
        System.out.println(trie.countWordsStartingWith("app"));

    }
}

class TrieNode {
    TrieNode[] children;
    int count;
    int pCount;

    public TrieNode() {
        children = new TrieNode[26];
        count = 0;
        pCount = 0;
    }
}

File: DSA/Python/Questions/Trie/ImplementTrie2.py
"""
# Question: Implement Trie II (Prefix Tree)
# Link: https://leetcode.com/problems/implement-trie-ii-prefix-tree/

# Implement a trie with word count tracking

# Time Complexity:
# - insert: O(n)
# - countWordsEqualTo: O(n)
# - countWordsStartingWith: O(n)
# - erase: O(n)

# Space Complexity: O(n)

# Algorithm:
# 1. Use TrieNode with children array
# 2. Track word count and prefix count
# 3. Implement insert, count, and erase operations
# 4. Maintain accurate counts during operations

# Key Components:
# - TrieNode class implementation
# - Word and prefix count tracking
# - Search functionality
"""


class TrieNode:
    def __init__(self):
        self.children = [None] * 26
        self.count = 0  # Count of words ending here
        self.prefix = 0  # Count of words using this prefix


class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        node = self.root
        for c in word:
            idx = ord(c) - ord("a")
            if not node.children[idx]:
                node.children[idx] = TrieNode()
            node = node.children[idx]
            node.prefix += 1
        node.count += 1

    def count_words_equal_to(self, word: str) -> int:
        node = self._search(word)
        return node.count if node else 0

    def count_words_starting_with(self, prefix: str) -> int:
        node = self._search(prefix)
        return node.prefix if node else 0

    def erase(self, word: str) -> None:
        node = self.root
        for c in word:
            idx = ord(c) - ord("a")
            node = node.children[idx]
            node.prefix -= 1
        node.count -= 1

    def _search(self, s: str) -> TrieNode:
        node = self.root
        for c in s:
            idx = ord(c) - ord("a")
            if not node.children[idx]:
                return None
            node = node.children[idx]
        return node


def main():
    trie = Trie()
    trie.insert("apple")
    trie.insert("apple")
    print(trie.count_words_equal_to("apple"))
    print(trie.count_words_starting_with("app"))
    trie.erase("apple")
    print(trie.count_words_equal_to("apple"))


if __name__ == "__main__":
    main()
